File: backend/app/schemas/risk_scoring.py
"""
F5 — Risk Scoring: Pydantic v2 request / response schemas.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


# ── Enums ──────────────────────────────────────────────────────────────────

class RiskLevel(str, Enum):
    LOW = "low"          # 0.0 – 0.3
    MEDIUM = "medium"    # 0.3 – 0.6
    HIGH = "high"        # 0.6 – 0.8
    CRITICAL = "critical"  # 0.8 – 1.0


class UpdateTrigger(str, Enum):
    NEW_TRANSACTION = "new_transaction"
    ALERT_RESOLUTION = "alert_resolution"
    WATCHLIST_MATCH = "watchlist_match"


# ── Component breakdown (F5.3) ──────────────────────────────────────────────

class ComponentScores(BaseModel):
    ml_score: float = Field(..., ge=0.0, le=1.0, description="ML ensemble fraud probability")
    rule_score: float = Field(..., ge=0.0, le=1.0, description="Rules engine trigger score")
    velocity_score: float = Field(..., ge=0.0, le=1.0, description="Transaction velocity anomaly score")
    behavioral_score: float = Field(..., ge=0.0, le=1.0, description="Behavioral deviation score")
    network_score: float = Field(..., ge=0.0, le=1.0, description="Network / fraud-ring score")


# ── Core risk score payload ──────────────────────────────────────────────────

class RiskScoreBase(BaseModel):
    entity_id: str
    overall_score: float = Field(..., ge=0.0, le=1.0)
    risk_level: RiskLevel
    component_scores: ComponentScores
    risk_factors: List[str] = Field(default_factory=list)
    model_version: str
    explanation: Optional[str] = None
    transaction_id: Optional[str] = None


class RiskScoreResponse(RiskScoreBase):
    id: str
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Entity risk profile (F5.1 / F5.3 / F5.4) ────────────────────────────────

class EntityRiskProfile(BaseModel):
    entity_id: str
    current_score: float = Field(..., ge=0.0, le=1.0)
    risk_level: RiskLevel
    component_scores: ComponentScores
    risk_factors: List[str]
    model_version: str
    explanation: Optional[str] = None
    last_updated: datetime
    score_trend: List[float] = Field(
        default_factory=list,
        description="Last 30 overall scores in chronological order",
    )


# ── Risk score history entry (F5.4) ──────────────────────────────────────────

class RiskScoreHistoryEntry(BaseModel):
    id: str
    overall_score: float
    risk_level: RiskLevel
    component_scores: ComponentScores
    risk_factors: List[str]
    transaction_id: Optional[str]
    created_at: datetime


class RiskScoreHistory(BaseModel):
    entity_id: str
    history: List[RiskScoreHistoryEntry]
    total: int


# ── On-demand calculate request / response (F5.1) ───────────────────────────

class CalculateRiskRequest(BaseModel):
    entity_id: str
    transaction_data: Dict = Field(
        ...,
        description="Raw transaction payload forwarded to ML pipeline",
    )
    watchlist_match: bool = Field(False, description="Whether entity has a watchlist hit")
    triggered_rules: List[str] = Field(default_factory=list)

    @field_validator("entity_id")
    @classmethod
    def entity_id_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("entity_id must not be empty")
        return v


class CalculateRiskResponse(BaseModel):
    entity_id: str
    overall_score: float
    risk_level: RiskLevel
    component_scores: ComponentScores
    risk_factors: List[str]
    explanation: str
    model_version: str
    record_id: str


# ── Risk distribution (F5.5) ─────────────────────────────────────────────────

class RiskTierCount(BaseModel):
    risk_level: RiskLevel
    count: int
    percentage: float


class RiskDistribution(BaseModel):
    total_entities: int
    tiers: List[RiskTierCount]
    thresholds: Dict[str, float] = Field(
        default_factory=lambda: {
            "low_max": 0.3,
            "medium_max": 0.6,
            "high_max": 0.8,
            "critical_max": 1.0,
        }
    )


# ── Top-N leaderboard (F5.6) ─────────────────────────────────────────────────

class TopRiskEntity(BaseModel):
    rank: int
    entity_id: str
    overall_score: float
    risk_level: RiskLevel
    risk_factors: List[str]
    last_updated: datetime


class TopRiskEntities(BaseModel):
    entities: List[TopRiskEntity]
    total_returned: int


# ── Auto-update trigger (F5.7) ───────────────────────────────────────────────

class AutoUpdateRequest(BaseModel):
    entity_id: str
    trigger: UpdateTrigger
    context: Dict = Field(default_factory=dict)


class AutoUpdateResponse(BaseModel):
    entity_id: str
    trigger: UpdateTrigger
    previous_score: float
    new_score: float
    risk_level: RiskLevel
    updated: bool

File: frontend/src/types/risk-scoring.ts
// F5 — Risk Scoring: TypeScript types

export type RiskLevel = "low" | "medium" | "high" | "critical";

export type UpdateTrigger =
  | "new_transaction"
  | "alert_resolution"
  | "watchlist_match";

// F5.3 — Component score breakdown
export interface ComponentScores {
  ml_score: number;
  rule_score: number;
  velocity_score: number;
  behavioral_score: number;
  network_score: number;
}

// F5.1 / F5.3 — Entity risk profile
export interface EntityRiskProfile {
  entity_id: string;
  current_score: number;
  risk_level: RiskLevel;
  component_scores: ComponentScores;
  risk_factors: string[];
  model_version: string;
  explanation: string | null;
  last_updated: string;
  score_trend: number[];
}

// F5.4 — Risk score history entry
export interface RiskScoreHistoryEntry {
  id: string;
  overall_score: number;
  risk_level: RiskLevel;
  component_scores: ComponentScores;
  risk_factors: string[];
  transaction_id: string | null;
  created_at: string;
}

export interface RiskScoreHistory {
  entity_id: string;
  history: RiskScoreHistoryEntry[];
  total: number;
}

// F5.1 — On-demand calculate
export interface CalculateRiskRequest {
  entity_id: string;
  transaction_data: Record<string, unknown>;
  watchlist_match?: boolean;
  triggered_rules?: string[];
}

export interface CalculateRiskResponse {
  entity_id: string;
  overall_score: number;
  risk_level: RiskLevel;
  component_scores: ComponentScores;
  risk_factors: string[];
  explanation: string;
  model_version: string;
  record_id: string;
}

// F5.5 — Risk distribution
export interface RiskTierCount {
  risk_level: RiskLevel;
  count: number;
  percentage: number;
}

export interface RiskDistribution {
  total_entities: number;
  tiers: RiskTierCount[];
  thresholds: Record<string, number>;
}

// F5.6 — Top-N leaderboard
export interface TopRiskEntity {
  rank: number;
  entity_id: string;
  overall_score: number;
  risk_level: RiskLevel;
  risk_factors: string[];
  last_updated: string;
}

export interface TopRiskEntities {
  entities: TopRiskEntity[];
  total_returned: number;
}

// F5.7 — Auto-update
export interface AutoUpdateRequest {
  entity_id: string;
  trigger: UpdateTrigger;
  context?: Record<string, unknown>;
}

export interface AutoUpdateResponse {
  entity_id: string;
  trigger: UpdateTrigger;
  previous_score: number;
  new_score: number;
  risk_level: RiskLevel;
  updated: boolean;
}

// UI helpers
export const RISK_LEVEL_CONFIG: Record<
  RiskLevel,
  { label: string; color: string; bg: string; border: string }
> = {
  low: {
    label: "Low",
    color: "text-green-700",
    bg: "bg-green-100",
    border: "border-green-300",
  },
  medium: {
    label: "Medium",
    color: "text-yellow-700",
    bg: "bg-yellow-100",
    border: "border-yellow-300",
  },
  high: {
    label: "High",
    color: "text-orange-700",
    bg: "bg-orange-100",
    border: "border-orange-300",
  },
  critical: {
    label: "Critical",
    color: "text-red-700",
    bg: "bg-red-100",
    border: "border-red-300",
  },
};

export const RISK_CHART_COLORS: Record<RiskLevel, string> = {
  low: "#22c55e",
  medium: "#eab308",
  high: "#f97316",
  critical: "#ef4444",
};
